apis/infrastructure/v1beta1/bootstrapkubeconfig_webhook.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package v1beta1
+
+import (
+	b64 "encoding/base64"
+	"encoding/pem"
+	"net/url"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var bootstrapkubeconfiglog = logf.Log.WithName("bootstrapkubeconfig-resource")
+
+// APIServerURLScheme is the url scheme for the APIServer
+const APIServerURLScheme = "https"
+
+func (r *BootstrapKubeconfig) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/validate-infrastructure-cluster-x-k8s-io-v1beta1-bootstrapkubeconfig,mutating=false,failurePolicy=fail,sideEffects=None,groups=infrastructure.cluster.x-k8s.io,resources=bootstrapkubeconfigs,verbs=create;update,versions=v1beta1,name=vbootstrapkubeconfig.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &BootstrapKubeconfig{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *BootstrapKubeconfig) ValidateCreate() error {
+	bootstrapkubeconfiglog.Info("validate create", "name", r.Name)
+
+	if err := r.validateAPIServer(); err != nil {
+		return err
+	}
+
+	if err := r.validateCAData(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *BootstrapKubeconfig) ValidateUpdate(old runtime.Object) error {
+	bootstrapkubeconfiglog.Info("validate update", "name", r.Name)
+
+	if err := r.validateAPIServer(); err != nil {
+		return err
+	}
+
+	if err := r.validateCAData(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *BootstrapKubeconfig) ValidateDelete() error {
+	bootstrapkubeconfiglog.Info("validate delete", "name", r.Name)
+
+	// TODO(user): fill in your validation logic upon object update.
+	return nil
+}
+
+func (r *BootstrapKubeconfig) validateAPIServer() error {
+	if r.Spec.APIServer == "" {
+		return field.Invalid(field.NewPath("spec").Child("apiserver"), r.Spec.APIServer, "APIServer field cannot be empty")
+	}
+
+	parsedURL, err := url.Parse(r.Spec.APIServer)
+	if err != nil || parsedURL.Host == "" || parsedURL.Scheme != APIServerURLScheme || parsedURL.Port() == "" {
```

Comment:
Would you like to separate it out in separate function something like `func IsURLValid()` as this contains multiple checks